filters/f_hwtransfer.c

```diff
@@ -94,10 +94,12 @@ static bool select_format(struct priv *p, int input_fmt,
     if (!input_fmt)
         return false;
 
-    // If the input format is a hw format, then we shouldn't be doing this
-    // format selection here at all.
+    // If the input format is a hw format, then we won't be doing any sort of
+    // conversion. Just assume that it will pass-through successfully.
```

Comment:
different angle of looking at this: should hwupload or hwdownload even be present if the format is not to be changed?
who is inserting them?

Reply:
It's actually not inserting a real filter (or at least it shouldn't be, but I'm not running the code in this exact configuration). What's actually going on here is it's executing all the logic that's required to prove that it doesn't need to do an upload or a hw->hw conversion. When the dust settles, it doesn't actually instantiate a filter.

So by assuming this pass-through just works, it sets things up so that the later code sees that no conversion is required and skips the filter.

The basic code flow here is that f_outputchain always instantiates f_autoconvert which always instantiates the hwupload instance, but hwupload is actually a wrapper and depending on what hwupload's init code does, it either creates a filter or skips it. Then f_autoconvert attaches the filter (if it was instantiated) to the filter chain.

Reply:
I also am bit concerned about this. I didn't follow the code fully. But it seems like we are short circuiting hw_input format to hw_output format, without actually checking if output is supported. It skips the logic of the conversion code, but the comment doesn't explain why this is acceptable. And as I understand this is only acceptable for rpi formats because they are mislabeled. I think this should be made explicit.

Also I'm not certain if the actual dummy filter is still not created, because we are already deep in the format selection, but like I said I didn't have time to review all the code, so I might be talking completely wrong.

Reply:
In this case, we will definitely not instantiate a filter because `src_is_same_hw` is true and there is no conversion filter for drm. So it will go down a path that doesn't create and return the filter.

Format selection happens before all that, because format selection is required to decide if the filter is needed.

All that said, I probably should revert the part where I appended the hwaccel format to the upload format list. As the select_format logic change means it just assumes an hwaccel format can be passed through, it doesn't need to be in the upload list.

Finally, yes - this path it is taking is absolutely just assuming things will work with no evidence, but we need to remember that was what the code did before I made all these conversion filter changes, so it's no worse than it was before, and as there is no actual conversion ability for drm, it's not as if there's an alternative. The vo will error out if it's impossible (as with the P030 example in the linked Issue)

Reply:
> The basic code flow here is that f_outputchain always instantiates f_autoconvert which always instantiates the hwupload instance, but hwupload is actually a wrapper and depending on what hwupload's init code does, it either creates a filter or skips it. Then f_autoconvert attaches the filter (if it was instantiated) to the filter chain.

okay that sounds fine.